src/record_syscall.cc

```diff
@@ -2957,6 +2958,30 @@ static pid_t do_detach_teleport(RecordTask *t)
   return new_tid;
 }
 
+template <typename Arch>
+static Switchable did_emulate_read(int syscallno, RecordTask* t,
+                                   const std::vector<FileMonitor::Range>& ranges,
+                                   uint64_t result,
+                                   TaskSyscallState& syscall_state)
+{
+  syscall_state.emulate_result(result);
+  record_ranges(t, ranges, result);
+  if (syscallno == Arch::pread64 || syscallno == Arch::preadv || result <= 0) {
+    // Don't perform this syscall.
+    Registers r = t->regs();
+    r.set_arg1(-1);
+    t->set_regs(r);
+  } else {
+    // Turn this into an lseek to emulate the advance of the fd ptr
+    Registers r = t->regs();
+    r.set_original_syscallno(Arch::lseek);
+    r.set_arg2(result);
+    r.set_arg3(SEEK_CUR);
+    t->set_regs(r);
```

Comment:
Seems like there is a potential issue if the real file is shorter than the emulated data, we might not be able to read all the emulated data, right?

Reply:
I think lseek should be able to set the file pointer to an arbitrary offset as long as SEEK_DATA and SEEK_HOLE are not specified, so I think this should work fine. Quick testing seems to agree with me:
```
keno@arctic4:~/rr-clean-build$ ./bin/rr record --num-cores=10 cat /sys/devices/system/cpu/online
rr: Saving execution to trace directory `/home/keno/.local/share/rr/cat-21'.
0,15,17,20,24,30,32,35,37,39
keno@arctic4:~/rr-clean-build$  cat /sys/devices/system/cpu/online
0-39
```
I think if the lseek hadn't worked, it would have repeated the first 5 bytes over and over.